seismicpro/src/seismic_batch.py

```diff
@@ -693,15 +706,45 @@ def drop_zero_traces(self, index, src, num_zero, **kwargs):
         -------
             : SeismicBatch
             Batch without dropped traces.
+
+        Raises
+        ------
+        ValueError : if `src` has no sorting and batch index is FieldIndex
+        ValueError : if any component in batch has sorting different from `src`
+
+        Note
+        ----
+        This action creates new instance of SeismicBatch with new index
+        instance.
         """
         _ = kwargs
+        sorting = self.meta[src]['sorting']
+        if sorting is None and isinstance(self.index, FieldIndex):
+            raise ValueError('traces in `{}` component should be sorted '
+                             'before dropping zero traces'.format(src))
+
+        has_same_sorting = all([self.meta[comp]['sorting'] == sorting for comp in self.components])
+        if not has_same_sorting:
+            raise ValueError('all components in batch should have same sorting')
+
         pos = self.get_pos(None, src, index)
         traces = getattr(self, src)[pos]
         mask = list()
-        for _, trace in enumerate(traces != 0):
-            diff_zeros = np.diff(np.append(np.where(trace)[0], len(trace)))
-            mask.append(False if len(diff_zeros) == 0 else np.max(diff_zeros) < num_zero)
-        return mask
+        for trace in traces:
+            nonzero_indices = np.nonzero(trace)[0]
+            # add -1 and len(trace) indices to count leading and trailing zero sequences
+            nonzero_indices = np.concatenate(([-1], nonzero_indices, [len(trace)]))
+            zero_seqs = np.diff(nonzero_indices) - 1
+            mask.append(np.max(zero_seqs) < num_zero)
+        mask = np.array(mask)
+
+        for comp in self.components:
+            getattr(self, comp)[pos] = getattr(self, comp)[pos][mask]
+
+        cols = [(INDEX_UID, src), (sorting, '')]
+        sorted_index_df = self.index.get_df(index)[cols].sort_values(sorting)
```

Comment:
```suggestion
        sorted_index_df = self.index.get_df([index])
```
In case `TraceIndex` you want the resulted slice to be `DataFrame`, not `Series`

Reply:
Indeed, for `TraceIndex` the logic should be different